src/test/java/com/liferay/faces/util/cache/CacheTest.java

```diff
@@ -0,0 +1,319 @@
+		testConcurrentCache(cache2, testErrors, testFailures);
+
+		final Cache<String, String> cache3 = cacheFactoryImpl.getConcurrentCache(defaultInitialCacheCapacity, 10);
+		testConcurrentMaxCacheCapacityCache(cache3, 10, testErrors, testFailures);
+	}
+
+	@Test
+	public void runConcurrentCacheTest() throws Exception {
+
+		CacheFactoryImpl cacheFactoryImpl = new CacheFactoryImpl();
+		Cache<String, String> cache1 = cacheFactoryImpl.getConcurrentCache();
+		testCache(cache1, 1000);
+
+		final Cache<String, String> cache2 = cacheFactoryImpl.getConcurrentCache();
+		final Queue<AssertionError> testFailures = new ConcurrentLinkedQueue<AssertionError>();
+		final Queue<Throwable> testErrors = new ConcurrentLinkedQueue<Throwable>();
+		testConcurrentCache(cache2, testErrors, testFailures);
+	}
+
+	private void errorOrFailTestIfNecessary(final Queue<Throwable> testErrors, final Queue<AssertionError> testFailures)
+		throws AssertionError, Exception {
+
+		for (Throwable testError : testErrors) {
+			logger.error("", testError);
+		}
+
+		for (AssertionError testFailure : testFailures) {
+			logger.error("", testFailure);
+		}
+
+		int testErrorsCapacity = testErrors.size();
+
+		if (testErrorsCapacity > 0) {
+			throw new Exception(testErrorsCapacity + " threads threw an error during test execution.");
+		}
+
+		int testFailuresCapacity = testFailures.size();
+
+		if (testFailuresCapacity > 0) {
+			throw new AssertionError(testFailuresCapacity + " threads reported a failure during test execution.");
+		}
+	}
+
+	private void testCache(Cache cache, int iterations) {
+
+		for (int i = 0; i < iterations; i++) {
+
+			String key = "key" + i;
+			String value = "value" + i;
+			Assert.assertNull(cache.get(key));
+			Assert.assertEquals(value, cache.put(key, value));
+			Assert.assertEquals(value, cache.putIfAbsent(key, "different" + value));
+			Assert.assertEquals(value, cache.get(key));
+			Assert.assertEquals("different" + value, cache.put(key, "different" + value));
+			Assert.assertEquals("different" + value, cache.get(key));
+
+			if ((i % 2) == 1) {
+
+				Assert.assertEquals("different" + value, cache.remove(key));
+				Assert.assertNull(cache.get(key));
+			}
+		}
+
+		Assert.assertEquals(iterations / 2, cache.keySet().size());
+	}
+
+	private void testConcurrentCache(Cache<String, String> cache, Queue<Throwable> testErrors,
+		Queue<AssertionError> testFailures) throws Exception {
+
+		Thread[] threads = new Thread[100];
+		Random random = new Random(System.currentTimeMillis());
+
+		for (int i = 0; i < threads.length; i++) {
+			threads[i] = new TestConcurrentCacheThread(cache, testFailures, testErrors, "key" + (i % 25), "value" + i,
+					random.nextInt(25));
+		}
+
+		for (Thread thread : threads) {
+			thread.start();
+		}
+
+		errorOrFailTestIfNecessary(testErrors, testFailures);
+	}
+
+	private void testConcurrentMaxCacheCapacityCache(final Cache cache, final int maxCacheCapacity,
```

Comment:
@stiemannkj1 IntelliJ reports many warnings because `Cache` is not parameterized. Please change to `Cache<String,String>` to eliminate these warnings.